### ISSUE TITLE
How to add a new download?

### ISSUE THREAD

**Issue by bzbug** [zsugabubus/aria2t#4]
Didn't find that in the man page.

Will there be adding of new downloads, as well as full functionality (like hand-picking certain files when downloading via BitTorrent), so that `aria2t` would be full interface for `aria2c` and so that it could replace GUI software such as `qBittorrent`?

**Comment by zsugabubus** [comment]
> ...so that aria2t would be full interface for aria2c and so that it could replace GUI software such as qBittorrent?

That's certainly possible. A little backstory: my dad was a long-time qBittorrent (previously Deluge) but some irritating bugs have popped up and it was pretty slooow. Back that time, I also had a hard time finding a suitable Torrent client, so that's how `aria2c` has born.

> Didn't find that in the man page.

Manual maybe a little bit cryptic so sorry for that. My English is not that shiny so if you find some places where you would reword something, please feel free to open an issue/PR.

> How to add a new download?

Basically, you can use three different keys for it: <kbd>a</kbd>, <kbd>A</kbd>, <kbd>+</kbd>. They do exactly the same, except they execute different actions/little helper programs, so you can easily define multiple preprogrammed ways for how to do it. By default (when no executable is placed under .config/aria2t/actions/<pressed-key>) you are simply given a file to edit (using editor specified by `$VISUAL`/`$EDITOR` environment variables).

The file you have to produce has a matching structure to aria2c's session file, namely:

* One download per line.
* When a line starts with space then this will be an option for the download above it.

Example:
```
https://site/a.pdf
    pause=true
      dir=/path/to/download/dir
/path/to/file1.torrent
/path/to/file2.torrent
  pause=false
```

Of course nobody wants to edit these files by hand every time, so that's when you can take advantage of the previously mentioned helper programs that will generate the content of this file for you.

Place an executable under .config/aria2t/actions/A:

```sh
#!/bin/sh
for f in ~/Downloads/*.torrent # Gather files.
do
  printf '%s\n' "$f"           # Specify location of the download.
  # Followed by some options...
  echo ' dir=/my/custom/dir'   # In case you would like to download to other location than the globally specified one.
  echo ' pause=true'           # Add in paused state.
done >"$2"                     # Write everything into the necessary file. aria2t will read it.

# In case you would like to further edit the generated files. Maybe you want to specify further options.
$EDITOR "$2"
```

Use URL from the clipboard: Place a script under .config/aria2t/actions/+.

```sh
#!/bin/sh
xclip -o >"$2"
```

> like hand-picking certain files when downloading via BitTorrent

Thats a bit more trickier.
1. Switch to files view. (E.g. <kbd>Right</kbd>)
2. Look at the files and remember their number. (Or remember just one and repeat the following steps multiple times.)
3. Press <kbd>i</kbd> to show options for the download.
4. Delete all lines from the file. (Important because these options may overwrite global options, so later changing a global option will not be used for this download.)
5. Type "select-file=1-4,7" (For the exact syntax see the `man aria2c` and search for "--select-file").
6. Save file and close the editor.

---

For more sophisticated scripts please see [my configuration](https://github.com/zsugabubus/dotfiles/tree/master/.config/aria2t/actions). In case you are interested, I also have some scripts for maintaining download directory and toggling tags on files.)

**Comment by bzbug** [comment]
The hand-picking process seems... not very in the spirit of UI. I would expect some interactive list where you could toggle 'checkboxes' or whatever to enable or disable certain files for download. What's the point of TUI if you still have to look up, remember and then write indexes instead of enabling/disabling things you see right in front of you with hotkeys?

Also, is it possible to create a separate view for new downloads where some options would be part of UI like in `qBitTorrent`? There, when you add a new torrent, the new 'view' asks you where to download, what files to pick, if you want to change the structure of the directory, if you want to download files consecutively - things like that. It doesn't make sense to edit a session file and write common options (such as `pause`) by hand in cases when the automation script does not apply, and it also doesn't make sense to write file paths by hand in your text editor when you have TUI file managers which can choose a directory automatically for you.

What I'm saying is, although _technically_ this all can be implemented through `actions` (and with the help of ugly workarounds in case of separate 'view' for new downloads; ugly because the view belongs _inside_ the `aria2` TUI, not outside it), it doesn't make sense to do it that way when the very program's goal is to create TUI that would get rid of redundant typing and remembering. So that's why it doesn't seem to be complete - it basically suggests that users should write other parts of TUI themselves for `aria2` TUI to be usable.

**Comment by zsugabubus** [comment]
First of all, thank you very much for the critic.

> The hand-picking process seems... not very in the spirit of UI. I would expect some interactive list where you could toggle 'checkboxes' or whatever to enable or disable certain files for download. What's the point of TUI if you still have to look up, remember and then write indexes instead of enabling/disabling things you see right in front of you with hotkeys?

Yeah, you are _absolutely_ right. Me neither likes the current behavior, but you know it was... ehhh... good for me... nobody complained so I didn't spend time with that.

Some good news is that this morning I have made some modifications that make this thing more comfortable. I will push changes soon but I would like to add batch operations (taking common actions on several downloads at once) too, that is also my old dream, to see how all these fit together.

> Also, is it possible to create a separate view for new downloads where some options would be part of UI like in qBitTorrent? 

I can imagine such thing that newly added downloads gets filtered (you only see them) and you can execute batch operations on them.

> There, when you add a new torrent, the new 'view' asks you where to download

(X) What about running your file manager at the very top of your action "A" and using that to populate `dir=` options?

> what files to pick

I think above mentioned batch stuff will make this possible.

> It doesn't make sense to edit a session file and write common options (such as pause) by hand in cases when the automation script does not apply

Why? How better could you do that? I mean you just need to edit a file in a fully fledged text editor. You can quickly add/copy/remove lines, copy/delete options you need. How could that be made more efficient?

> it also doesn't make sense to write file paths by hand in your text editor when you have TUI file managers which can choose a directory automatically for you

If you frequently need to download to different places, I think (X) would be a good solution. No?

Or even better, if you can script your file manager, it could also generate the file that can be fed into `aria2t`.

> with the help of ugly workarounds in case of separate 'view' for new downloads; ugly because the view belongs inside the aria2 TUI, not outside it

Could you explain the latter part in other words?

> it basically suggests that users should write other parts of TUI themselves for aria2 TUI to be usable.

Think about actions as configuration options.

These actions can be shared among users, even included in `aria2t`'s repository so they are discoverable and can be quickly integrated into everybody's environment by picking the necessary files.

Returning a bit to (X), personally I would not like that the program to ask me every time. It's a checkbox in `qBitTorrent` and commenting a line in `aria2t`/using different action file.
